### ISSUE TITLE
IO that actually does... io

### ISSUE THREAD

**Issue by benjamingr** [cwmyers/monet.js#25]
Hey, you have an IO type, nice. 

Wouldn't it be nice if it would be able to actually perform IO (like wait for the user to enter a string, make a web request, read a file in node and so on)?

All it currently does is emulate `unsafePerformIO` - wouldn't it be useful for it to do IO?

**Comment by cwmyers** [comment]
Hehe. The point of the IO type is to wrap IO side-effects to make them referentially transparent. Monet is a monad library not an effects library! Hopefully this explains it more: http://cwmyers.github.io/monet.js/#io



—

**Comment by benjamingr** [comment]
@cwmyers here is an IO monad implementation in JS https://github.com/fantasyland/fantasy-promises fwiw

**Comment by cwmyers** [comment]
It sure is. :-)



—

**Comment by benjamingr** [comment]
Basically, you'd use promises - a promise `then` signature (if you disregard exceptions and returning unwrapped  values) is a `Promise a -> (a -> Promise b) -> Promise b` - sounds familiar :)?

**Comment by cwmyers** [comment]
Fantasyland is awesome. Feel free to use it to make your effects referentially transparent. I do have something in my backlog to be fantasyland compliant. I had made slightly different design choices to Brian. But like I said I've got no plans to make monet an IO effects library. Maybe one day I'll spin off a monet-effects project like scalaz did with scalaz-streams. 




Thanks for you comments. 



—

**Comment by benjamingr** [comment]
Right, but what's stopping you from doing IO in a referentially transparent
way in monet? Right now how do you perform async actions?

**Comment by scottburch** [comment]
@benjamingr I am new to this library and am still figuring out proper ways to use Monads.  This is a POC that I put together to see if I can use Monet for async operations.  Does this work for you?  Let me know if I am way off.  It may be a bit of an abuse of Reader.  Not really sure.

    function readFromServer(conf, fn) {
        var m = fn(Reader(function(v) {return v}));
    
        // simulate async response
        setTimeout(function() {
            m.run('data received using '+conf);
        },100);
    }


    function upperCase(string) {
        return string.toUpperCase();
    }

    readFromServer('some conf', function(reader) {
        return reader.map(upperCase).map(console.log.bind(console));
    });

**Comment by benjamingr** [comment]
@scottburch it works for me, but it's not monadic - instead it uses a callback. 

A monadic way to do this would be something like:

    do 
        data <- readFromServer 'some conf'
        let upper = upperCase data
        print upper

Or, in JavaScript I'd assume that something like:

     readFromServer('some conf').map(upperCase).bind(print)

Note the function is not passed into the `readFromServer` function, rather the operation (bind) is applied to the monad creating the `M a -> (a -> M b) -> M b` bind signature

**Comment by scottburch** [comment]
@benjamingr Yes, I would also prefer that structure as well.  But, I could not figure out a way to get it to work with the current library.  What the lib might need is a Promise monad.  I might look into adding one.

**Comment by scottburch** [comment]
@benjamingr Here is some rough code I threw together to test how a Promise monad might work.  If you could be so kind as to let me know what you think I would appreciate it.  I'm returning an Either monad so that I can handle either success or failure.

    Promise = function() {
        var mapFn;
        var returnedPromise; 
        return {
            map: function(fn) {
                mapFn = fn;
                return returnedPromise = Promise();
            },
            resolve: function(data) {
                var e = typeof (data.unit) === 'function' ? data : Either.Right(data);
                returnedPromise && returnedPromise.resolve(e.map(mapFn));
            },
            reject: function(data) {
                var e = typeof (data.unit) === 'function' ? data : Either.Left(data);
                returnedPromise && returnedPromise.reject(e.map(mapFn));
            }
        }
    };

    function readFromServer(conf) {
        var p = Promise();
        setTimeout(function() {
            p.resolve('here is the data: '+conf);
        });
        return p;
    }

    function readFromServerFail(conf) {
        var p = Promise();
        setTimeout(function() {
            p.reject('error');
        });
    }


    // Code that does the work
    readFromServer('my conf').map(upperCase).map(console.log.bind(console));

    function upperCase(string) {
        return string.toUpperCase();
    }

**Comment by scottburch** [comment]
Created a small monet-promise library for anyone interested.  I have not used it yet in the real world, but will today and will update soon if necessary.

https://github.com/scottburch/monet-promise

**Comment by benjamingr** [comment]
This is actually `io`, also, your monet-promise is sort of like a promise but it's no monad. Where's the bind?

**Comment by scottburch** [comment]
If you look in the test, bind is in there.  I'm slowly building it.  If
it works for you, use it.  If not, no biggie. :)  Any comments are
welcome though.
I did do a promise thing with IO, but the Reader worked better since
run took a parameter.  With the monet-promise I am using Either so that
I can have a pass/fail thing.  It is mostly about referential
transparency and being able to chain functions using bind/map.
Scott

**Comment by paulvictor** [comment]
Don't know if people have figured this out somehow, I had done a little bit of experimentation myself and found Observables(from rxjs) to serve the purpose, so created a ObservableT (using Monad transformers) to create a higher order monad which can do actual IO, here is a gist in case someone's interested https://gist.github.com/paulvictor/87e8390c55b3dafcef2b . The existing implementation of MonadT i think has some problem wrt to the definition and implementation of bind(aka flatMap). See https://en.wikibooks.org/wiki/Haskell/Monad_transformers for more on Monad Transformers.
Paul

**Comment by cwmyers** [comment]
I'm not happy with the implementation of the monad transformer which is why
it is undocumented. It will be removed in a future version.

**Comment by paulvictor** [comment]
Agreed about the implementation, but monad transformers are quite helpful, would be helpful just to define the ```map```,``` ap```, ```of``` and other combinators, allowing the actual implementation to provide an implementation for ```bind```.

**Comment by cwmyers** [comment]
Indeed they are. But having a single monadT is wrong. Monad transformers
need awareness over what they are transforming. MonadT was an experiment.

In future there will be specific transformers such as MaybeT, EitherT etc.

:-)
